### ISSUE TITLE
PCR thermocycler sequence

### ISSUE THREAD

**Issue by TheChymera** [biopython/biopython#498]
Is there any function in biopython to help me determine what temperature and duration to set for my PCR thermocycler steps given a certain amplicon and certain primers?

If not, I might be happy to try and contribute this, but looking it up just gave me lots of voodoo. Obviously this also depends on stuff like which polymerase you are using etc. but there should be some sort of equation nontheless.

**Comment by peterjc** [comment]
There are functions in ``Bio.SeqUtils`` for some relevant calculations, see eg https://github.com/biopython/biopython/blob/master/Bio/SeqUtils/MeltingTemp.py

Ideally there would be a published method/algorithm we could implement in Biopython - I suggest you sign up to the Biopython mailing list and raise this issue there, perhaps someone in our community  knows more:
http://biopython.org/wiki/Mailing_lists
http://lists.open-bio.org/mailman/listinfo/biopython/

**Comment by peterjc** [comment]
Thanks for posting on the mailing list - hopefully someone will reply over the weekend or next week...
http://mailman.open-bio.org/pipermail/biopython/2015-March/015592.html

**Comment by campenr** [comment]
I feel this is a hard problem to solve and that it at best would give a rough estimate for a starting point as there are so many variables. Melting temperature is easy as it purely depends on the length/base content of your sequence, and it has a nice clean solution.

Annealing temperature for a PCR reaction is a more complicated issue as it depends on both the melting temperature of your DNA as well as the polymerase you are using, but also lab specific variables like tube wall thickness, how well sealed the tubes are, do you use DMSO or similar and how much, machines ramping time. With these lab specific variables even websites like http://tmcalculator.neb.com/ at best give you an approximate starting point.

In terms of an equation for amplification time it would be `amp_time = seconds_per_kb * length_in_kb` but does that deserve its own BioPython function? Also this is again only a starting point, you may need to up the time if the desired product is greater than a certain length (polymerase specific) and decrease if you're getting unwanted products that are longer than your desired product.

**Comment by TheChymera** [comment]
@campenr I agree, though in my experience with PCR, having some sane and robust guidelines is more important than getting all the details 100% correct. I have used tubes from at least 3 different suppliers just within the last 4 months, and I never noticed any unexpected changes in my PCR results.

I guess we could make some sane assumptions, and some insane (ideal) ones:

* DMSO 0%
* Ramping time: 0s 
* Amplification speed: 60s/kbp

Then the function could take differing numbers from optional arguments, so that if the defaults prove unsatisfactory, or the user knows he is using a really nonstandard setup, or really wants to get it totally perfect.

I think it is safe to exclude parameters such as "thickness of tube walls" or "well-sealdness of the tube" unless someone actually runs into issues with these things.


I am more unsure about the technical details of implementing this. Should it be a class? shloud it be a function? What format should we export in?

I am currently using `.csv` files formatted like this (though it seems to me that this maynot be too well fitted for more complex protocols):

```
step,"temperature [°C]","duration [s]","gradient [°C]","ramp rate [°C/s]","return to step",iterations
0,92,180,,,,
1,95,60,,,,
2,94,30,,,,
3,58,60,,,,
4,72,120,,,2,5
5,94,30,,,,
6,25,120,,,,
7,72,240,,0.4,,
8,94,30,,,,
9,58,60,,,,
10,72,120,,,,
11,94,30,,,,
12,58,60,,,,
13,72,120,,,,
14,94,30,,,,
15,42,60,,,,
16,72,120,,,8,15
17,72,300,,,,
18,4,inf,,,,
```

**Comment by campenr** [comment]
This is definitely doable if you want to have something that can automate the math that is used to estimate annealing temp and extension time, e.g.

    annealing_temp = average_Tm +/- polymerase_specific_variable
    extension_time = kilobases * rate_of_polymerase

My caution comes from the fact that defining other parameters for PCR is not an exact science (annealing time, number of cycles, etc). There are norms and best practices for setting these parameters, but I wonder if a problem such as this with such a fuzzy solution fits into BioPython? Or would it be better suited simply built on top of BioPython but not a part of it?

**Comment by campenr** [comment]
Regarding the actual program to do this. I think that you would want a function, say `create_protocol` that takes the variables as keyword arguments, e.g. polymerase_rate, primer1_Tm, primer2_Tm, annealing_time with default arguments for the variables. I like the idea of storing in csv like you show.

**Comment by TheChymera** [comment]
As I said, we can (and should) also add calculations for DMSO content and ramp rate (in case the user specifies hem).

I think this has a place in Biopython since it automates a procedure which most people do by hand time and time and time again, manually.

Not least of all it's either here or nowhere. I can write a separate package for it, which I might use a number of times, but nobody is going to bother installing a separate package just for this (as they would not for many other single Biopython function)

Lastly, I think it would be better to design this as a class, so that I can use notations like:
`my_pcr_protocol.extensiontime` or `my_pcr_protocol.Ta` or `py_pcr_protocol.csv` to access different things about it. What do you think?

**Comment by campenr** [comment]
Yeah maybe a class would be the way to go. My though is perhaps a ThermalCycle class that contains all the attributes for the thermal cycle, as well as methods for writing them to file, and parsing those files. As well as perhaps a method for creating visuals for the cycle, who knows. Is there a better forum to discuss this? I don't want to clog up the issue tracker for BioPython.

**Comment by TheChymera** [comment]
I usually prefer discussing new features on the github tracker, it's always easier to link it to pull requests and link pull requests back to it, add tags etc.

Though maybe we should brainstorm for ideas on the mailing list before we set the outline of how this could be done? Would you like to post there? (I already posted a while back, with little avail).
